event/src/unvalidated/builder.rs

```diff
@@ -203,9 +212,151 @@ impl<D> DataBuilder<DataBuilderWithData<D>> {
+
+impl TimeBuilder<TimeBuilderWithId> {
+    /// Specify details about the time transaction.
+    pub fn with_tx(
+        self,
+        chain_id: String,
+        tx_hash: Cid,
+        tx_type: String,
+    ) -> TimeBuilder<TimeBuilderWithTx> {
+        TimeBuilder {
+            state: TimeBuilderWithTx {
+                id: self.state.id,
+                chain_id,
+                tx_hash,
+                tx_type,
+            },
+        }
+    }
+}
+
+/// State with the transaction details.
+pub struct TimeBuilderWithTx {
+    id: Cid,
+    chain_id: String,
+    tx_hash: Cid,
+    tx_type: String,
+}
+impl TimeBuilderState for TimeBuilderWithTx {}
+
+impl TimeBuilder<TimeBuilderWithTx> {
+    /// Specify the root of the proof.
+    /// The index is an index into the edge that should be followed.
+    pub fn with_root(self, index: usize, edge: Ipld) -> TimeBuilder<TimeBuilderWithRoot> {
+        TimeBuilder {
+            state: TimeBuilderWithRoot {
+                id: self.state.id,
+                chain_id: self.state.chain_id,
+                tx_hash: self.state.tx_hash,
+                tx_type: self.state.tx_type,
+                edges: vec![(index, edge)],
+            },
+        }
+    }
+}
+
+/// State with the proof root added.
+/// More edges may be added.
+pub struct TimeBuilderWithRoot {
+    id: Cid,
+    chain_id: String,
+    tx_hash: Cid,
+    tx_type: String,
+    edges: Vec<(usize, Ipld)>,
+}
+impl TimeBuilderState for TimeBuilderWithRoot {}
+impl TimeBuilder<TimeBuilderWithRoot> {
+    /// Specify an additional edge of the proof.
+    /// The index is an index into the edge that should be followed.
+    /// The last edge must index to a Cid of the previous event.
+    pub fn with_edge(mut self, index: usize, edge: Ipld) -> Self {
+        self.state.edges.push((index, edge));
+        self
+    }
+    /// Build the [`unvalidated::TimeEvent`].
+    /// Errors if the proof edges and indexes are not valid.
+    pub fn build(self) -> anyhow::Result<unvalidated::TimeEvent> {
+        let path = self
+            .state
+            .edges
+            .iter()
+            .map(|(index, _edge)| index.to_string())
+            .collect::<Vec<_>>()
+            .join("/");
+        let (_index, root) = self
+            .state
+            .edges
+            .first()
+            .expect("should always be at least one edge");
+        let (leaf_index, leaf_edge) = self
+            .state
+            .edges
+            .iter()
+            .last()
+            .expect("should always be at least one edge");
+        let prev = leaf_edge
+            .get(*leaf_index)?
+            .ok_or_else(|| anyhow!("leaf index should always exist"))?;
+        let prev = match prev {
+            Ipld::Link(prev) => *prev,
+            _ => bail!("leaf indexed value should always be a Cid"),
+        };
+        let root_bytes = serde_ipld_dagcbor::to_vec(root)?;
+        let root_cid = cid_from_dag_cbor(&root_bytes);
```

Comment:
I don't understand how this is working.

The root is an IPLD array with two elements, each the CID of the two child nodes in the tree.  But the proof only has one of the child nodes in the path.  So then how can you compute the root CID without knowing the CID of the other child of the root that *isn't* a part of the path for this TimeEvent?


EDIT: Nevermind I understand now after looking at how this code is used in the unit test.  See my comment above, but I found the names and comments on the functions for building up the path confusing and lead to me not understanding how they were actually used until I saw the unit test.